File: backend/app/models/api/secrets.py
"""Schémas Pydantic pour les endpoints /v1/wallets/{id}/secrets/* — LOT_05."""
from __future__ import annotations

import base64
import re
from uuid import UUID

from pydantic import BaseModel, field_validator

# Limite : 5 MB après décodage base64
_MAX_VALUE_BYTES = 5 * 1024 * 1024

# Regex env-var-safe
_NAME_RE = re.compile(r"^[A-Za-z0-9_.\-]+$")


# ─── Requêtes ─────────────────────────────────────────────────────────────────


class SecretCreateRequest(BaseModel):
    """Corps de POST /v1/wallets/{id}/secrets."""

    name: str
    description: str | None = None
    tags: list[str] = []
    encrypted_value: str  # base64

    @field_validator("name")
    @classmethod
    def _name_valid(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("name must not be empty")
        if len(stripped) > 256:
            raise ValueError("name must not exceed 256 characters")
        if not _NAME_RE.match(stripped):
            raise ValueError(
                "name must match ^[A-Za-z0-9_.-]+ (env-var-safe characters only)"
            )
        return stripped

    @field_validator("description")
    @classmethod
    def _desc_valid(cls, v: str | None) -> str | None:
        if v is not None and len(v) > 1000:
            raise ValueError("description must not exceed 1000 characters")
        return v

    @field_validator("tags", mode="before")
    @classmethod
    def _tags_normalize(cls, v: object) -> list[str]:
        if not isinstance(v, list):
            raise ValueError("tags must be a list")
        return [str(t).strip().lower() for t in v if str(t).strip()]

    @field_validator("encrypted_value")
    @classmethod
    def _value_valid(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("encrypted_value must not be empty")
        try:
            raw = base64.b64decode(v)
        except Exception as exc:
            raise ValueError("encrypted_value must be valid base64") from exc
        if len(raw) > _MAX_VALUE_BYTES:
            raise ValueError(
                f"encrypted_value exceeds max size of {_MAX_VALUE_BYTES} bytes after decoding"
            )
        return v


class SecretPutRequest(BaseModel):
    """Corps de PUT /v1/wallets/{id}/secrets/{name} — mise à jour valeur."""

    encrypted_value: str  # base64

    @field_validator("encrypted_value")
    @classmethod
    def _value_valid(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("encrypted_value must not be empty")
        try:
            raw = base64.b64decode(v)
        except Exception as exc:
            raise ValueError("encrypted_value must be valid base64") from exc
        if len(raw) > _MAX_VALUE_BYTES:
            raise ValueError(
                f"encrypted_value exceeds max size of {_MAX_VALUE_BYTES} bytes after decoding"
            )
        return v


class SecretPatchRequest(BaseModel):
    """Corps de PATCH /v1/wallets/{id}/secrets/{name} — métadonnées uniquement."""

    description: str | None = None
    tags: list[str] | None = None

    @field_validator("description")
    @classmethod
    def _desc_valid(cls, v: str | None) -> str | None:
        if v is not None and len(v) > 1000:
            raise ValueError("description must not exceed 1000 characters")
        return v

    @field_validator("tags", mode="before")
    @classmethod
    def _tags_normalize(cls, v: object) -> list[str] | None:
        if v is None:
            return None
        if not isinstance(v, list):
            raise ValueError("tags must be a list")
        return [str(t).strip().lower() for t in v if str(t).strip()]


# ─── Réponses ─────────────────────────────────────────────────────────────────


class _CallerRef(BaseModel):
    type: str
    id: UUID


class SecretListItem(BaseModel):
    """Item dans la liste GET /v1/wallets/{id}/secrets — sans encrypted_value."""

    id: UUID
    name: str
    description: str | None
    tags: list[str]
    is_placeholder: bool
    generation_version: int
    linked_secret_id: UUID | None
    created_at: str  # ISO 8601
    updated_at: str  # ISO 8601
    created_by: _CallerRef | None
    updated_by: _CallerRef | None


class SecretListResponse(BaseModel):
    """Réponse de GET /v1/wallets/{id}/secrets."""

    secrets: list[SecretListItem]
    next_cursor: str | None


class SecretDetailResponse(BaseModel):
    """Réponse de GET /v1/wallets/{id}/secrets/{name}."""

    id: UUID
    name: str
    encrypted_value: str  # base64
    encrypted_wallet_key: str  # base64 — clé du caller depuis wallet_grants
    description: str | None
    tags: list[str]
    is_placeholder: bool
    generation_version: int


class SecretCreateResponse(BaseModel):
    """Réponse de POST /v1/wallets/{id}/secrets."""

    secret_id: UUID


class SecretPutResponse(BaseModel):
    """Réponse de PUT /v1/wallets/{id}/secrets/{name}."""

    generation_version: int

File: frontend/src/schemas/secrets.ts
/**
 * Zod schemas for /v1/wallets/{id}/secrets/* endpoints.
 * Mirror of backend app/models/api/secrets.py.
 */
import { z } from 'zod'

const NAME_RE = /^[A-Za-z0-9_.-]+$/
const DESC_MAX = 1000

export const SecretCreateRequestSchema = z.object({
  name: z
    .string()
    .min(1)
    .max(256)
    .regex(NAME_RE, 'name must match ^[A-Za-z0-9_.-]+ (env-var-safe)'),
  description: z.string().max(DESC_MAX).nullable().optional(),
  tags: z.array(z.string()).default([]),
  encrypted_value: z.string().min(1),
})

export const SecretPutRequestSchema = z.object({
  encrypted_value: z.string().min(1),
})

export const SecretPatchRequestSchema = z.object({
  description: z.string().max(DESC_MAX).nullable().optional(),
  tags: z.array(z.string()).nullable().optional(),
})

const CallerRefSchema = z.object({
  type: z.string(),
  id: z.string().uuid(),
})

export const SecretListItemSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  description: z.string().nullable(),
  tags: z.array(z.string()),
  is_placeholder: z.boolean(),
  generation_version: z.number().int(),
  linked_secret_id: z.string().uuid().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
  created_by: CallerRefSchema.nullable(),
  updated_by: CallerRefSchema.nullable(),
})

export const SecretListResponseSchema = z.object({
  secrets: z.array(SecretListItemSchema),
  next_cursor: z.string().nullable(),
})

export const SecretDetailResponseSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  encrypted_value: z.string(),
  encrypted_wallet_key: z.string(),
  description: z.string().nullable(),
  tags: z.array(z.string()),
  is_placeholder: z.boolean(),
  generation_version: z.number().int(),
})

export const SecretCreateResponseSchema = z.object({
  secret_id: z.string().uuid(),
})

export const PopulateRequestSchema = z.object({
  encrypted_value: z.string().min(1),
})

export const PlaceholderCreateRequestSchema = z.object({
  name: z
    .string()
    .min(1)
    .max(256)
    .regex(NAME_RE, 'name must match ^[A-Za-z0-9_.-]+'),
  description: z.string().max(DESC_MAX).nullable().optional(),
  tags: z.array(z.string()).default([]),
  generation_descriptor: z.record(z.unknown()),
  linked_secret_id: z.string().uuid().nullable().optional(),
})

export type SecretListItem = z.infer<typeof SecretListItemSchema>
export type SecretDetailResponse = z.infer<typeof SecretDetailResponseSchema>

